File: avalon/src/java/org/apache/avalon/framework/configuration/DefaultConfigurationSerializer.java
/*

 ============================================================================
                   The Apache Software License, Version 1.1
 ============================================================================
 
 Copyright (C) @year@ The Apache Software Foundation. All rights reserved.
 
 Redistribution and use in source and binary forms, with or without modifica-
 tion, are permitted provided that the following conditions are met:
 
 1. Redistributions of  source code must  retain the above copyright  notice,
    this list of conditions and the following disclaimer.
 
 2. Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution.
 
 3. The end-user documentation included with the redistribution, if any, must
    include  the following  acknowledgment:  "This product includes  software
    developed  by the  Apache Software Foundation  (http://www.apache.org/)."
    Alternately, this  acknowledgment may  appear in the software itself,  if
    and wherever such third-party acknowledgments normally appear.
 
 4. The names "Jakarta", "Apache Avalon", "Avalon Excalibur", "Avalon
    Framework" and "Apache Software Foundation"  must not be used to endorse
    or promote products derived  from this  software without  prior written
    permission. For written permission, please contact apache@apache.org.
 
 5. Products  derived from this software may not  be called "Apache", nor may
    "Apache" appear  in their name,  without prior written permission  of the
    Apache Software Foundation.
 
 THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED WARRANTIES,
 INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 FITNESS  FOR A PARTICULAR  PURPOSE ARE  DISCLAIMED.  IN NO  EVENT SHALL  THE
 APACHE SOFTWARE  FOUNDATION  OR ITS CONTRIBUTORS  BE LIABLE FOR  ANY DIRECT,
 INDIRECT, INCIDENTAL, SPECIAL,  EXEMPLARY, OR CONSEQUENTIAL  DAMAGES (INCLU-
 DING, BUT NOT LIMITED TO, PROCUREMENT  OF SUBSTITUTE GOODS OR SERVICES; LOSS
 OF USE, DATA, OR  PROFITS; OR BUSINESS  INTERRUPTION)  HOWEVER CAUSED AND ON
 ANY  THEORY OF LIABILITY,  WHETHER  IN CONTRACT,  STRICT LIABILITY,  OR TORT
 (INCLUDING  NEGLIGENCE OR  OTHERWISE) ARISING IN  ANY WAY OUT OF THE  USE OF
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 This software  consists of voluntary contributions made  by many individuals
 on  behalf of the Apache Software  Foundation and was  originally created by
 Stefano Mazzocchi  <stefano@apache.org>. For more  information on the Apache 
 Software Foundation, please see <http://www.apache.org/>.
 
*/
package org.apache.avalon.framework.configuration;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.net.URL;
import java.util.Properties;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.sax.SAXTransformerFactory;
import javax.xml.transform.sax.TransformerHandler;
import javax.xml.transform.stream.StreamResult;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.AttributesImpl;
import org.xml.sax.helpers.NamespaceSupport;

/**
 * A ConfigurationSerializer serializes configurations via SAX2 compliant parser.
 *
 * @author <a href="mailto:bloritsch@apache.org">Berin Loritsch</a>
 * @version 1.0
 */
public class DefaultConfigurationSerializer
{
    private SAXTransformerFactory m_tfactory;
    private TransformerHandler    m_handler;
    private OutputStream          m_out;
    private Properties            m_format = new Properties();
    private NamespaceSupport      m_namespaceSupport = new NamespaceSupport();

    /**
     * Build a ConfigurationSerializer
     */
    public DefaultConfigurationSerializer()
    {
        getTransformerFactory();
    }

    /**
     * Sets the Serializer's use of indentation.  This will cause linefeeds to be added
     *  after each element, but it does not add any indentation via spaces.
     * @param indent a <code>boolean</code> value
     */
    public void setIndent( boolean indent )
    {
        if ( indent )
        {
            m_format.put( OutputKeys.INDENT, "yes" );
        }
        else
        {
            m_format.put( OutputKeys.INDENT, "no" );
        }
    }

    /**
     * Internally set the output strream we will be using.
     * @param out an <code>OutputStream</code> value
     */
    protected void setOutputStream( final OutputStream out )
    {
        try
        {
            m_out = out;
            m_handler = getTransformerFactory().newTransformerHandler();
            m_format.put(OutputKeys.METHOD,"xml");
            m_handler.setResult(new StreamResult(out));
            m_handler.getTransformer().setOutputProperties( m_format );
        }
        catch( final Exception e )
        {
            throw new RuntimeException( e.toString() );
        }
    }

    /**
     * Get the SAXTransformerFactory so we can get a serializer without being
     * tied to one vendor.
     * @return a <code>SAXTransformerFactory</code> value
     */
    protected SAXTransformerFactory getTransformerFactory()
    {
        if(m_tfactory == null)
        {
            m_tfactory = (SAXTransformerFactory) TransformerFactory.newInstance();
        }

        return m_tfactory;
    }

    /**
     * Start the serialization process.  The output stream <strong>must</strong>
     * be set before calling this method.
     * @param source a <code>Configuration</code> value
     * @throws SAXException if an error occurs
     * @throws ConfigurationException if an error occurs
     */
    protected void serialize( final Configuration source )
        throws SAXException, ConfigurationException
    {
        m_namespaceSupport.reset();
        m_handler.startDocument();
        serializeElement(source);
        m_handler.endDocument();
    }

    /**
     * Serialize each Configuration element.  This method is called recursively.
     * @param element a <code>Configuration</code> value
     * @throws SAXException if an error occurs
     * @throws ConfigurationException if an error occurs
     */
    protected void serializeElement( final Configuration element )
        throws SAXException, ConfigurationException
    {
        m_namespaceSupport.pushContext();

        AttributesImpl attr = new AttributesImpl();
        String[] attrNames = element.getAttributeNames();

        if (null != attrNames)
        {
            for (int i = 0; i < attrNames.length; i++)
            {
                attr.addAttribute("", // namespace URI
                                  attrNames[i], // local name
                                  attrNames[i], // qName
                                  "CDATA",  // type
                                   element.getAttribute(attrNames[i], "") // value
                                 );
            }
        }

        final String nsURI = element.getNamespace();
        String nsPrefix = "";

        if ( element instanceof AbstractConfiguration )
        {
            nsPrefix = ((AbstractConfiguration) element).getPrefix();
        }
        // nsPrefix is guaranteed to be non-null at this point.

        boolean nsWasDeclared = false;

        final String existingURI = m_namespaceSupport.getURI( nsPrefix );

        // ie, there is no existing URI declared for this prefix or we're
        // remapping the prefix to a different URI
        if ( existingURI == null || !existingURI.equals( nsURI ) )
        {
            nsWasDeclared = true;
            if (nsPrefix.equals("") && nsURI.equals(""))
            {
                // implicit mapping; don't need to declare
            }
            else if (nsPrefix.equals(""))
            {
                // (re)declare the default namespace
                attr.addAttribute("", "xmlns", "xmlns", "CDATA", nsURI);
            }
            else
            {
                // (re)declare a mapping from nsPrefix to nsURI
                attr.addAttribute("", "xmlns:"+nsPrefix, "xmlns:"+nsPrefix, "CDATA", nsURI);
            }
            m_handler.startPrefixMapping( nsPrefix, nsURI );
            m_namespaceSupport.declarePrefix( nsPrefix, nsURI );
        }

        String localName = element.getName();
        String qName = element.getName();
        if ( nsPrefix == null || nsPrefix.length() == 0 )
        {
            qName = localName;
        }
        else
        {
            qName = nsPrefix + ":" + localName;
        }

        m_handler.startElement(nsURI, localName, qName, attr);

        String value = element.getValue(null);

        if (null == value)
        {
            Configuration[] children = element.getChildren();

            for (int i = 0; i < children.length; i++)
            {
                serializeElement(children[i]);
            }
        }
        else
        {
            m_handler.characters(value.toCharArray(), 0, value.length());
        }

        m_handler.endElement(nsURI, localName, qName);

        if ( nsWasDeclared )
        {
            m_handler.endPrefixMapping( nsPrefix );
        }

        m_namespaceSupport.popContext();
    }

    /**
     * Serialize the configuration object to a file using a filename.
     * @param filename a <code>String</code> value
     * @param source a <code>Configuration</code> value
     * @throws SAXException if an error occurs
     * @throws IOException if an error occurs
     * @throws ConfigurationException if an error occurs
     */
    public void serializeToFile( final String filename, final Configuration source )
        throws SAXException, IOException, ConfigurationException
    {
        serializeToFile( new File( filename ), source );
    }

    /**
     * Serialize the configuration object to a file using a File object.
     * @param file a <code>File</code> value
     * @param source a <code>Configuration</code> value
     * @throws SAXException if an error occurs
     * @throws IOException if an error occurs
     * @throws ConfigurationException if an error occurs
     */
    public void serializeToFile( final File file, final Configuration source )
        throws SAXException, IOException, ConfigurationException
    {
        serialize( new FileOutputStream( file ), source );
    }

    /**
     * Serialize the configuration object to an output stream.
     * @param outputStream an <code>OutputStream</code> value
     * @param source a <code>Configuration</code> value
     * @throws SAXException if an error occurs
     * @throws IOException if an error occurs
     * @throws ConfigurationException if an error occurs
     */
    public void serialize( final OutputStream outputStream, final Configuration source )
        throws SAXException, IOException, ConfigurationException
    {
        synchronized(this)
        {
            setOutputStream( outputStream );
            serialize( source );
        }
    }

    /**
     * Serialize the configuration object to an output stream derived from an
     * URI.  The URI must be resolveable by the <code>java.net.URL</code> object.
     * @param uri a <code>String</code> value
     * @param source a <code>Configuration</code> value
     * @throws SAXException if an error occurs
     * @throws IOException if an error occurs
     * @throws ConfigurationException if an error occurs
     */
    public void serialize( final String uri, final Configuration source )
        throws SAXException, IOException, ConfigurationException
    {
        serialize( new URL( uri ).openConnection().getOutputStream(), source );
    }
}

File: avalon-sandbox/avalon-net/csframework/AvalonFramework/DefaultConfigurationSerializer.cs
// ============================================================================
//                   The Apache Software License, Version 1.1
// ============================================================================
// 
// Copyright (C) 2002-2003 The Apache Software Foundation. All rights reserved.
// 
// Redistribution and use in source and binary forms, with or without modifica-
// tion, are permitted provided that the following conditions are met:
// 
// 1. Redistributions of  source code must  retain the above copyright  notice,
//    this list of conditions and the following disclaimer.
// 
// 2. Redistributions in binary form must reproduce the above copyright notice,
//    this list of conditions and the following disclaimer in the documentation
//    and/or other materials provided with the distribution.
// 
// 3. The end-user documentation included with the redistribution, if any, must
//    include  the following  acknowledgment:  "This product includes  software
//    developed  by the  Apache Software Foundation  (http://www.apache.org/)."
//    Alternately, this  acknowledgment may  appear in the software itself,  if
//    and wherever such third-party acknowledgments normally appear.
// 
// 4. The names "Jakarta", "Avalon", "Excalibur" and "Apache Software Foundation"  
//    must not be used to endorse or promote products derived from this  software 
//    without  prior written permission. For written permission, please contact 
//    apache@apache.org.
// 
// 5. Products  derived from this software may not  be called "Apache", nor may
//    "Apache" appear  in their name,  without prior written permission  of the
//    Apache Software Foundation.
// 
// THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED WARRANTIES,
// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS  FOR A PARTICULAR  PURPOSE ARE  DISCLAIMED.  IN NO  EVENT SHALL  THE
// APACHE SOFTWARE  FOUNDATION  OR ITS CONTRIBUTORS  BE LIABLE FOR  ANY DIRECT,
// INDIRECT, INCIDENTAL, SPECIAL,  EXEMPLARY, OR CONSEQUENTIAL  DAMAGES (INCLU-
// DING, BUT NOT LIMITED TO, PROCUREMENT  OF SUBSTITUTE GOODS OR SERVICES; LOSS
// OF USE, DATA, OR  PROFITS; OR BUSINESS  INTERRUPTION)  HOWEVER CAUSED AND ON
// ANY  THEORY OF LIABILITY,  WHETHER  IN CONTRACT,  STRICT LIABILITY,  OR TORT
// (INCLUDING  NEGLIGENCE OR  OTHERWISE) ARISING IN  ANY WAY OUT OF THE  USE OF
// THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
// 
// This software  consists of voluntary contributions made  by many individuals
// on  behalf of the Apache Software  Foundation. For more  information on the 
// Apache Software Foundation, please see <http://www.apache.org/>.
// ============================================================================

namespace Apache.Avalon.Framework
{
	using System;
	using System.Collections;
	using System.IO;
	using System.Xml;

	/// <summary>
	/// A Serializer/Deserializer of a <see cref="DefaultConfiguration"/>.
	/// </summary>
	public class DefaultConfigurationSerializer
	{
		/// <summary>
		/// Makes a serialization of a <see cref="DefaultConfiguration"/> instance.
		/// </summary>
		/// <param name="filename">
		/// The File name where <see cref="DefaultConfiguration"/> instance will be
		/// serialized to.
		/// </param>
		/// <param name="configuration">A <see cref="DefaultConfiguration"/> instance to serialize</param>
		public static void Serialize(string filename, DefaultConfiguration configuration)
		{
			XmlTextWriter writer = new XmlTextWriter(new StreamWriter(filename));

			//Use indentation for readability.
			writer.Formatting = Formatting.Indented;
			writer.Indentation = 4;
			
			writer.WriteStartDocument(true);

			Serialize(writer, configuration);				

			writer.WriteEndDocument(); 
			writer.Close(); 
		}

		/// <summary>
		/// Makes a serialization of a <see cref="DefaultConfiguration"/> instance.
		/// </summary>
		/// <param name="writer"></param>
		/// <param name="configuration">A <see cref="DefaultConfiguration"/> instance to serialize.</param>
		public static void Serialize(XmlWriter writer, DefaultConfiguration configuration)
		{
			// serialize the configuration
			writer.WriteStartElement(configuration.Prefix, configuration.Name, configuration.Namespace); 
			
			// attribute serialization
			foreach (DictionaryEntry attr in configuration.Attributes) 
			{
				writer.WriteAttributeString(attr.Key.ToString(), attr.Value.ToString()); 
			}  

			if (configuration.Value != null)
			{
				writer.WriteString(configuration.Value.ToString());
			}

			// child serialization
			foreach(IConfiguration child in configuration.Children)
			{
				Serialize(writer, (DefaultConfiguration) child);
			}

			writer.WriteEndElement();
		}

		/// <summary>
		/// Makes a deserialization of a <see cref="DefaultConfiguration"/> instance.
		/// </summary>
		/// <param name="fileName">The Name of the file, containing the XML document to deserialize.</param>
		/// <returns>A Deserialized <see cref="DefaultConfiguration"/> instance.</returns>
		public static DefaultConfiguration Deserialize(string fileName)
		{
			DefaultConfiguration configuration = null;

			XmlDocument document = new XmlDocument();
			document.Load(fileName);
			
			XmlNode root = document.DocumentElement; 

			if (root != null)
			{
				configuration = Deserialize(root, null);
			}

			return configuration;
		}

		/// <summary>
		/// Makes a deserialization of <see cref="XmlNode"/> instance.
		/// </summary>
		/// <param name="node">The Node to deserialize.</param>
		/// <param name="parent">A Deserialized <see cref="DefaultConfiguration"/> parent instance.</param>
		/// <returns>A Deserialized <see cref="DefaultConfiguration"/> instance.</returns>
		public static DefaultConfiguration Deserialize(XmlNode node, DefaultConfiguration parent)
		{
			// node deserialization
			DefaultConfiguration configuration = null;

			if ((node.NodeType  == XmlNodeType.CDATA) || (node.NodeType == XmlNodeType.Text))
			{
				if (parent != null)
				{
					parent.Value = string.Concat(parent.Value, node.Value);
				}
			}

			if ((node.NodeType  == XmlNodeType.Document) || (node.NodeType == XmlNodeType.Element))
			{
				configuration = new DefaultConfiguration(node.LocalName, "-", node.NamespaceURI, node.Prefix);
	
				// attribute deserialization
				if (node.Attributes != null)
				{
					foreach (XmlAttribute attr in node.Attributes)
					{
						if (string.Compare(attr.Prefix, string.Empty) == 0)
						{
							configuration.Attributes[attr.Name] = attr.Value;  
						}
					}
				}

				// child deserialization
				foreach (XmlNode child in node.ChildNodes)
				{
					DefaultConfiguration childConfiguration = Deserialize(child, configuration);
					
					if (childConfiguration != null)
					{
						configuration.Children.Add(childConfiguration); 
					}
				}
			}

			return configuration;
		}

		/// <summary>
		/// Makes a deserialization of <see cref="XmlNode"/> instance.
		/// </summary>
		/// <param name="node">The Node to deserialize.</param>
		/// <returns>A Deserialized <see cref="DefaultConfiguration"/> instance.</returns>
		public static DefaultConfiguration Deserialize(XmlNode node)
		{
			return Deserialize(node, null);
		}
	}
}
